File: python/VCFOrphan.py
from __future__ import annotations

# coding: utf-8
# VCFOrphan.py
# 孤立したサンプルを集めたVCF

from typing import List, Tuple

from VCF import VCFSmall
from VCFGeno import VCFGenoBase
from GenoRecord import GenoRecord
from VCFFamily import *
from OrphanImputer import *
from Genotype import Genotype


#################### VCFOrphan ####################

DP = List[Tuple[float, int]]	# (log of probability, prev h)
MIN_PROB = -1e300

class VCFOrphan(VCFGenoBase):
	def __init__(self, samples: list[str], records: list[GenoRecord],
						ref_haps: list[list[int]], map_: Map, vcf: VCFSmall):
		VCFGenoBase.__init__(self, samples, vcf)
		self.imputer = OrphanImputer(records, ref_haps, map_, 0.01)
		self.records: list[GenoRecord] = records
	
	def __len__(self) -> int:
		return len(self.records)
	
	def get_record(self, i: int) -> GenoRecord:
		return self.records[i]
	
	def get_samples(self) -> list[str]:
		return self.samples
	
	def impute(self) -> None:
		for i in range(self.num_samples()):
			self.imputer.impute(i)

File: src/VCFOrphan.cpp
#include <cmath>
#include "../include/VCFOrphan.h"
#include "../include/OrphanImputer.h"
#include "../include/common.h"

using namespace std;

VCFOrphan::VCFOrphan(const STRVEC& s, const std::vector<GenoRecord *>& rs,
						const std::vector<std::vector<int>>& ref_hs,
						const Map& map_, double w, const VCFSmall *vcf) :
				VCFGenoBase(s, vcf),
				records(rs),
				imputer(records, ref_hs, map_, w) { }

VCFOrphan::~VCFOrphan() {
	Common::delete_all(records);
}

void VCFOrphan::impute_each(size_t i) {
	imputer.impute(i);
}

void VCFOrphan::impute_small_in_thread(void *config) {
	auto	*c = static_cast<const ConfigThread *>(config);
	auto	*vcf = c->vcf;
	const size_t	n = vcf->num_samples();
	for(size_t i = c->first; i < n; i += c->num_threads) {
		vcf->impute_each(i);
	}
}

void VCFOrphan::impute(int T) {
	vector<ConfigThread *>	configs(T);
	for(int i = 0; i < T; ++i)
		configs[i] = new ConfigThread(i, T, this);
	
#ifndef DEBUG
	vector<pthread_t>	threads_t(T);
	for(int i = 0; i < T; ++i)
		pthread_create(&threads_t[i], NULL,
			(void *(*)(void *))&impute_small_in_thread, (void *)configs[i]);
	
	for(int i = 0; i < T; ++i)
		pthread_join(threads_t[i], NULL);
#else
	for(int i = 0; i < T; ++i)
		impute_small_in_thread(configs[i]);
#endif
	
	Common::delete_all(configs);
}
